lizrd/text/packers.py

```diff
@@ -206,3 +206,100 @@ def get_sample(self) -> LLMExample:
         calculate_loss = [1] * len(target_ids)
 
         return LLMExample(input_ids, target_ids, calculate_loss)
+
+
+class BlankPacker(
+    AbstractPacker,
+):
+    def __init__(
+        self,
+        sequence_length: int,
+        dataset: AbstractDataset,
+        tokenizer_maker: Callable[[], AbstractTokenizer],
+        n_blanks: int,
+        seed: Optional[int] = None,
+    ):
+        super().__init__(
+            sequence_length,
+            dataset,
+            tokenizer_maker,
+            seed=seed,
+        )
+        self.n_blanks = n_blanks
+        assert n_blanks > 0
+
+    def get_sample(self) -> LLMExample:
+        """
+        Sample examples from the dataset until we reach the desired sequence length.
+        """
+        eot_id = self.tokenizer.eot_id
+        blank_id = self.tokenizer.blank_id
+        assert eot_id is not None
+        assert blank_id is not None
+
+        buffer_input: List[int] = []
+        buffer_output: List[int] = []
+        blank_mask_buffer: List[int] = []
+        calculate_loss: List[int] = []
+        document_lengths: List[int] = []
+
+        # change it so the blank can come first in both example and sequence
+        while True:
+            document = self.dataset.get_document()
+            tokens = self.tokenizer.text_to_ids(document)
+            tokens.append(eot_id)
+
+            blank_insertion_point = self.py_rng.randint(1, len(tokens) - 1)
+            input_tokens = (
+                tokens[:blank_insertion_point]
+                + [blank_id] * self.n_blanks
+                + tokens[blank_insertion_point:]
+            )
+            output_tokens = (
+                tokens[:blank_insertion_point]
+                + [tokens[blank_insertion_point]] * self.n_blanks
+                + tokens[blank_insertion_point:]
+            )
+            blank_mask = (
+                [0] * blank_insertion_point
+                + [1] * self.n_blanks
+                + [0] * (len(tokens) - blank_insertion_point)
+            )
+            # for i in range(self.n_blanks + 1):
+            #     n_th_blank_mask[i].extend(
+            #         [0] * (blank_insertion_point + i - 1)
+            #         + ([1])
+            #         + [0] * (self.n_blanks - i + len(tokens) - blank_insertion_point)
+            #     )
+
+            # blank_insertion_point + i - 1 + 1 + self.n_blanks - i + len(tokens) - blank_insertion_point
+            # len(tokens) + blanks
+            buffer_input.extend(input_tokens)
+            buffer_output.extend(output_tokens)
+            blank_mask_buffer.extend(blank_mask)
+
+            document_lengths.append(len(tokens))
+            if (sum(document_lengths) - max(document_lengths)) > self.sequence_length:
```

Comment:
Jeśli byśmy w jakiś sposób liczyli sumę dokumentów nie licząc blanków, to mielibyśmy, chyba, te same dokumenty co baseline? do dyskusji